### PR TITLE
[PLUGIN-1754] Add Retry with dev.failsafe

### PR DESCRIPTION
**Pull request by psainics**
## Add Retry with dev.failsafe

Jira : [PLUGIN-1754](https://cdap.atlassian.net/browse/PLUGIN-1754)

### Description

When API Fails or rate limited, we want to retry after some wait, this PR adds `dev.failsafe` to handle the retry !
This is done by throwing a custom exception if the error is retryable.

### UI Field

- Modified `Ariba-batchsource.json`

### Docs

- No Changes made to docs.

### Code change

- Added `AribaRetryableException.java`
  -  Custom exception class if thrown the method will be retried. 
- Modified `AribaBatchSource.java`
- Modified `AribaInputFormat.java`
- Modified `AribaServices.java`
- Modified `AribaPluginConfig.java`
- Modified `AribaConnector.java`
- Modified `AribaConnectorConfig.java`

### Unit Tests

- Modified `AribaBatchSourceTest.java`
- Modified `AribaInputFormatTest.java`
- Modified `AribaRecordReaderTest.java`
- Modified `AribaServicesTest.java`
- Modified `PropertiesTest.java`

### CODE REVIEW

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java [hunk 1]
```diff
@@ -366,16 +381,8 @@ public List<AribaColumnMetadata> getMetadata(String accessToken, String template
...
     Response response;
-    do {
-      response = executeRequest(req);
-      if (response.code() == ResourceConstants.API_LIMIT_EXCEED) {
-        checkAndThrowException(response);
-        count++;
-      }
-    } while (response.code() == ResourceConstants.API_LIMIT_EXCEED && count <= MAX_RETRIES);
-
+    response = executeRequest(req);
```
Comment (anup-codex) [inline src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]: lets merge the line#384.

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java [hunk 2]
```diff
@@ -621,40 +624,74 @@ protected Request buildFetchRequest(URL endPoint, String accessToken) {
...
+   * @return Response
+   * @throws AribaException          AribaException
+   * @throws InterruptedException    InterruptedException
+   * @throws IOException             IOException
+   * @throws AribaRetryableException
+   */
+  public Response executeRetryableRequest(Request req)
+    throws AribaException, InterruptedException, IOException, AribaRetryableException {
+
+    LOG.debug("ARIBA URL: " + req.url());
```
Comment (anup-codex) [inline src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]: Update it to `Retryable Ariba URL:`

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java [hunk 3]
```diff
@@ -621,40 +624,74 @@ protected Request buildFetchRequest(URL endPoint, String accessToken) {
...
+        throw (InterruptedException) t;
+      } else if (t instanceof IOException) {
+        throw (IOException) t;
+      } else {
+        throw new RuntimeException(t);
+      }
+    }
+  }
+
+  /**
```
Comment (anup-codex) [inline src/main/java/io/cdap/plugin/ariba/source/AribaServices.java:667]: Add one line comment here.

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java [hunk 4]
```diff
@@ -621,40 +624,74 @@ protected Request buildFetchRequest(URL endPoint, String accessToken) {
...
+        String errMsg = jsonNode.get(ResourceConstants.MESSAGE).asText() != null
+          ? jsonNode.get(ResourceConstants.MESSAGE).asText() :
+          ResourceConstants.ERR_NOT_FOUND.getMsgForKey();
+        throw new AribaException(errMsg, response.code());
+      }
       checkAndThrowException(response);
       return response;
-    } else {
-      response = enhancedOkHttpClient.newCall(req).execute();
+    } catch (IOException e) {
```
Comment (anup-codex) [inline src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]: Lets remove this `catch` block and let it throw the original error to the caller.

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java [hunk 5]
```diff
@@ -665,15 +702,10 @@ public JsonNode fetchData(String jobId, String fileName)
     HttpUrl.Builder zipUrl = zipBuilder(jobId, fileName);
     Response zipResponse;
     do {
-      zipResponse = enhancedOkHttpClient
-        .newCall(fetchZipFileData(zipUrl.build().url(), getAccessToken())).execute();
-      if (zipResponse.code() == ResourceConstants.API_LIMIT_EXCEED) {
-        checkAndThrowException(zipResponse);
-      }
+      zipResponse = executeRequest(fetchZipFileData(zipUrl.build().url(), getAccessToken()));
     } while (zipResponse.code() == ResourceConstants.API_LIMIT_EXCEED);
```
Comment (anup-codex) [inline src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]: we can remove this line as `executeRequest` will perform the retry.

File: src/test/java/io/cdap/plugin/ariba/source/AribaBatchSourceTest.java [hunk 6]
```diff
@@ -295,19 +262,14 @@ public void testValidateAdvancedParametersError6() {
...
   public void testIsSchemaBuildRequired() {
-    pluginConfig = new AribaPluginConfig("referenceName",
-                                         "https://stackoverflow.com/questions/17225948/" +
-                                           "parsing-error-for-date-field{{{browser_user_agent}}}",
-                                         "prod", "realm",
-                                         "viewTemplateName", "clientId",
-                                         "clientSecret", "apiKey", "tokenURL",
-                                         "2022-01-28T10:05:02Z", "2023-01-28T10:05:02Z");
+    pluginConfig = pluginConfigBuilder
+      .baseURL("https://stackoverflow.com/questions/17225948/parsing-error-for-date-field{{{browser_user_agent}}}")
```
Comment (anup-codex) [inline src/test/java/io/cdap/plugin/ariba/source/AribaBatchSourceTest.java]: update this url to som dummy url.

Comment (psainics) [reply src/test/java/io/cdap/plugin/ariba/source/AribaBatchSourceTest.java]:
```java
@Test
  public void testIsSchemaBuildRequired() {
    pluginConfig = pluginConfigBuilder
      .baseURL("https://example.com/{{{browser_user_agent}}}")
      .build();
    Assert.assertTrue(pluginConfig.isSchemaBuildRequired());
  }
  ```

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java (hunk 5 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]:
```java
Response zipResponse = executeRequest(fetchZipFileData(zipUrl.build().url(), getAccessToken()));
```

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java (hunk 4 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]: Hmm, no point of try block too then ! 
Removed !

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java (hunk 3 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/ariba/source/AribaServices.java:667]: > This method is used to fetch the data from the Ariba API.

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java (hunk 2 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]:
```java
LOG.debug("Retryable Ariba URL: " + req.url());
```

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java [hunk 7]
```diff
@@ -663,17 +690,9 @@ public JsonNode fetchData(String jobId, String fileName)
 
     OkHttpClient enhancedOkHttpClient = getConfiguredClient().build();
```
Comment (anup-codex) [inline src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]: this line can be removed as it is not used.

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java [hunk 8]
```diff
@@ -723,10 +742,13 @@ public boolean isApiLimitExhausted(Response response) {
    * @throws InterruptedException
    */
   @VisibleForTesting
-  void checkAndThrowException(Response response) throws AribaException, InterruptedException {
+  void checkAndThrowException(Response response) throws AribaException, InterruptedException, AribaRetryableException {
     if (response.code() == HttpURLConnection.HTTP_BAD_REQUEST && !AribaUtil.isNullOrEmpty(response.message())) {
       throw new AribaException(response.message(), response.code());
     }
+    if (response.code() != HttpURLConnection.HTTP_OK && AribaUtil.isNullOrEmpty(response.message())) {
```
Comment (anup-codex) [inline src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]: remove this block as at the end it is taken care.

File: src/main/java/io/cdap/plugin/ariba/source/exception/AribaRetryableException.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
```
Comment (anup-codex) [inline src/main/java/io/cdap/plugin/ariba/source/exception/AribaRetryableException.java]: update it to current year.

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java (hunk 3 above)
Comment (anup-codex) [reply src/main/java/io/cdap/plugin/ariba/source/AribaServices.java:667]: > Calls given Ariba API.

**Review by anup-codex — APPROVED** [review]
LGTM, please fix the minor comments and raise the external PR.

File: src/main/java/io/cdap/plugin/ariba/source/exception/AribaRetryableException.java (hunk 9 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/ariba/source/exception/AribaRetryableException.java]: Updated !

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java (hunk 8 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]: Removed !

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java (hunk 7 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/ariba/source/AribaServices.java]: Removed !

File: src/main/java/io/cdap/plugin/ariba/source/AribaServices.java (hunk 3 above)
Comment (psainics) [reply src/main/java/io/cdap/plugin/ariba/source/AribaServices.java:667]: Fixed !

**Pull request merged**